authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,156 @@
 #  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  * See the License for the specific language governing permissions and
 #  * limitations under the License.
-
+import logging
 
 import ldap
 from flask_securest.authentication_providers.abstract_authentication_provider \
     import AbstractAuthenticationProvider
 from flask_securest import utils
+from flask_securest import constants
+from flask_securest.exceptions import AuthenticationException
+
+DN_ATTRIBUTE_NAME = 'DN'
+DEFAULT_AUTH_ATTR_NAME = 'uid'
+
+REQ_SEARCH_FIELDS = ('base_dn', 'admin_user_id', 'admin_password')
 
 
 class LDAPAuthenticationProvider(AbstractAuthenticationProvider):
-    def __init__(self, directory_url):
-        self.directory_url = directory_url
 
+    _logger = logging.getLogger(constants.FLASK_SECUREST_LOGGER_NAME)
+
+    def __init__(self, ldap_url, search_properties=None):
+        self.ldap_url = ldap_url
+        self.search_props = search_properties
+        self.validate_search_props()
+
+    def validate_search_props(self):
+        if self.search_props:
+            valid = all(test in self.search_props for test in
+                        REQ_SEARCH_FIELDS)
+            if not valid:
+                err = 'The configured search_properties are missing required '\
+                      'search fields. Search properties must hold values for '\
+                      'the following fields: \'{req_fields}\', got '\
+                      '\'{props_fields}\''\
+                    .format(req_fields=REQ_SEARCH_FIELDS,
+                            props_fields=self.search_props.keys())
+                self._logger.warning(err)
+                raise RuntimeError(err)
+
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.search_props:
+            user_dn = self.get_dn_by_user_attribute(who,
+                                                    ldap_url=self.ldap_url,
+                                                    **self.search_props)
+        else:
+            user_dn = who
+        # attempt to bind using the matching user ID/DN and provided password.
+        self.validate_bind(user_dn, password)
+        return who
+
+    @classmethod
+    def handle_authentication_error(cls, err, conn):
+        conn.unbind()
+        cls._logger.warning(err)
+        raise AuthenticationException(err)
+
+    def validate_bind(self, who, password):
```

Comment:
utility method, you can pass self.ldap_url as the first arg and then make it a classmethod